[site] Run RayJobs (#780)
kubernetes-sigs/kueue: site/content/en/docs/tasks/run_rayjobs.md

---
title: "Run A RayJob"

> date: 2022-05-18

Reviewer: ```suggestion
date: 2023-05-18
```

Author: done

Check [administer cluster quotas](/docs/tasks/administer_cluster_quotas) for details on the initial cluster setup.

Check [KubeRay](https://ray-project.github.io/kuberay/) documentation for installation and configuration details.

## RayJob definition

Besides the usual RayJob configuration while using Kueue take into consideration the following aspects:

### a. Queue selection

The target [local queue](/docs/concepts/local_queue) should be specified in the `metadata.labels` section of the RayJob configuration.

```yaml
metadata:
  labels:
    kueue.x-k8s.io/queue-name: user-queue
```

### b. Configure the resource needs

The resource needs of the workload can be configured in the `spec.rayClusterSpec`.

```yaml
    headGroupSpec:
      template:
        spec:
          containers:
            - resources:
                requests:
                  cpu: "1"
    workerGroupSpecs:
      - template:
          spec:
            containers:
              - resources:
                  requests:
                    cpu: "1"
```

### c. Limitations

- A Kueue managed RayJob cannot use an existing cluster.

> - The RayJob's cluster should be deleted at the end of its execution, `spec.ShutdownAfterJobFinishes` should be `true`.

Reviewer: ```suggestion
- The Ray cluster should be deleted at the end of its execution, `spec.ShutdownAfterJobFinishes` should be `true`.
```

Author: done

> - Because Kueue will reserve the RayJob resources, the cluster's `inTreeAutoscaling` should not be enabled.

Reviewer 2: ```suggestion
- Because Kueue will reserve the RayJob resources, `spec.rayClusterSpec.enableInTreeAutoscaling` should be `false`.
```

Reviewer: how about `will reserve resources for the Ray cluster`

Author: done

> - Because a Kueue workload can have a maximum of 8 PodSets, the maximum number of `workerGroupSpecs` is 7.

Reviewer 2: ```suggestion
- Because a Kueue workload can have a maximum of 8 PodSets, the maximum number of `spec.rayClusterSpec.workerGroupSpecs` is 7.
```

Author: done

## Sample RayJob

```yaml
apiVersion: ray.io/v1alpha1

> kind: RayJob

Reviewer 2: Is this inspired by https://github.com/ray-project/kuberay/blob/82c925bb3fe72e056a21521c3b6fff3b172c2817/ray-operator/config/samples/ray_v1alpha1_rayjob.yaml?

If so, it would be worth mentioning that as a comment.

Author: It kind of is, but the config map is added and it's `generateName` based  sou you can crete multiple jobs base in the same yaml.

metadata:
  generateName: sample-rayjob-
  labels:
    kueue.x-k8s.io/queue-name: user-queue
spec:

>   suspend: true

Reviewer 2: Do you prefer explicitly set `suspend=true`?

Author: no, is set by the default weeb-hook anyaway, removed

Check [administer cluster quotas](/docs/tasks/administer_cluster_quotas) for details on the initial cluster setup.

Check [KubeRay](https://ray-project.github.io/kuberay/) documentation for installation and configuration details.

## RayJob definition

Besides the usual RayJob configuration while using Kueue take into consideration the following aspects:

### a. Queue selection

The target [local queue](/docs/concepts/local_queue) should be specified in the `metadata.labels` section of the RayJob configuration.

```yaml
metadata:
  labels:
    kueue.x-k8s.io/queue-name: user-queue
```

### b. Configure the resource needs

The resource needs of the workload can be configured in the `spec.rayClusterSpec`.

```yaml
    headGroupSpec:
      template:
        spec:
          containers:
            - resources:
                requests:
                  cpu: "1"
    workerGroupSpecs:
      - template:
          spec:
            containers:
              - resources:
                  requests:
                    cpu: "1"
```

### c. Limitations

> - A Kueue managed RayJob cannot use an existing cluster.

Reviewer 3: ```suggestion
- A Kueue managed RayJob cannot use an existing RayCluster.
```

Maybe also add a link to what a Ray cluster is. Otherwise it can be confused with kubernetes cluster.

> - The RayCluster should be deleted at the end of the job execution, `spec.ShutdownAfterJobFinishes` should be `true`.

Reviewer 3: Should we do this in the webhook?

Author: It's a bit too "magical", since the Ray default is ~false.

- Because Kueue will reserve resources for the RayCluster, `spec.rayClusterSpec.enableInTreeAutoscaling` should be `false`.

> - Because a Kueue workload can have a maximum of 8 PodSets, the maximum number of `spec.rayClusterSpec.workerGroupSpecs` is 7.

Reviewer 3: Does RayJob have any validation for maximum number of `WorkerGroupSpecs`?

Author: no

## Sample RayJob

```yaml
apiVersion: ray.io/v1alpha1
kind: RayJob
metadata:
  generateName: sample-rayjob-
  labels:
    kueue.x-k8s.io/queue-name: user-queue
spec:
  shutdownAfterJobFinishes: true
  entrypoint: python /home/ray/samples/sample_code.py

>   runtimeEnv: ewogICAgInBpcCI6IFsKICAgICAgICAicmVxdWVzdHM9PTIuMjYuMCIsCiAgICAgICAgInBlbmR1bHVtPT0yLjEuMiIKICAgIF0sCiAgICAiZW52X3ZhcnMiOiB7ImNvdW50ZXJfbmFtZSI6ICJ0ZXN0X2NvdW50ZXIifQp9Cg==

Reviewer 3: is this something users need to provide?

Author: looks like it 

```json
{
    "pip": [
        "requests==2.26.0",
        "pendulum==2.1.2"
    ],
    "env_vars": {"counter_name": "test_counter"}
}
```
anyway is part of  RayJob official sample.

Reviewer 3: oh, this is a HEX yaml?

Odd
